src/mame/machine/dmac3.h

```diff
@@ -0,0 +1,193 @@
+// license:BSD-3-Clause
+// copyright-holders:Brice Onken,Tsubai Masanari
+// thanks-to:Patrick Mackinlay
+
+/*
+ * Sony CXD8403Q DMAC3 DMA controller
+ *
+ * Register definitions were derived from the NetBSD source code, copyright (c) 2000 Tsubai Masanari.
+ *
+ * References:
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/dmac3reg.h
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/dmac3var.h
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/dmac3.c
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/spifi.c
+ */
+
+#ifndef MAME_MACHINE_DMAC3_H
+#define MAME_MACHINE_DMAC3_H
+
+#pragma once
+
+class dmac3_device : public device_t
+{
+public:
+	dmac3_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock);
+
+	// DMAC3 has two controllers on-chip
+	// The 5000X uses controllers 0 and 1 for SPIFI/SCSI bus 0 and 1 respectively
+	enum dmac3_controller
+	{
+		CTRL0 = 0,
+		CTRL1 = 1,
+	};
+
+	// Address map setup
+	template <typename... T>
+	void set_apbus_address_translator(T &&...args) { m_apbus_virt_to_phys_callback.set(std::forward<T>(args)...); }
+	template <dmac3_controller controller>
+	void map(address_map &map)
```

Comment:
Please use `LlamaCase` for template parameter names (it seems you did for the other member functions, but not this one).